Fix: Backspace does not clear selected item in insert-completion

girishji opened the pull request:
**Fix:** Backspace does not clear selected item in insert-completion

**Problem:**
After selecting and inserting a completion item, pressing `<BS>` to delete part of the inserted text does **not** clear the selection.

* The completion item stays selected even though the text no longer matches.
* Pressing `<C-N>` moves to the next item instead of restarting from the first.
* If only one item is available, `<C-N>` does nothing (buggy).
* This differs from the behavior of other editors/IDEs.

<img width="237" height="177" alt="Screenshot 2025-09-10 at 8 40 11 AM" src="https://github.com/user-attachments/assets/03307c00-6d44-449e-8909-ae5df12f6d19" />


**Expected behavior:**
Once the inserted text no longer matches the selected item, it should be deselected.

**Notes:**
This bug is also present in `v9010754`, so it may have been longstanding.
Fixed buggy tests.

girishji committed:
Fix: Backspace does not clear selected item in insert-completion

**Fix:** Backspace does not clear selected item in insert-completion

**Problem:**
After selecting and inserting a completion item, pressing `<BS>` to
delete part of the inserted text does **not** clear the selection.

* The completion item stays selected even though the text no longer matches.
* Pressing `<C-N>` moves to the next item instead of restarting from the first.
* If only one item is available, `<C-N>` does nothing (buggy).
* This differs from the behavior of other editors/IDEs.

**Expected behavior:**
Once the inserted text no longer matches the selected item, it should be
deselected.

**Notes:**
This bug is also present in `v9010754`, so it may have been longstanding.

gcanat commented on src/insexpand.c:
```diff
@@ -2602,6 +2606,11 @@ ins_compl_addleader(int c)
 	    return;
 	}
 
+	// "preinsert" insertes the first match via compl_shown_match
```
small typo
```suggestion
	// "preinsert" inserts the first match via compl_shown_match
```

girishji committed:
Update src/insexpand.c

Co-authored-by: gcanat <72149218+gcanat@users.noreply.github.com>

girishji committed:
Apply this change only for autocompletion

This change is scoped only to autocomplete, so long-time Vim users can
continue with the current behavior if they prefer.

M  src/insexpand.c
M  src/testdir/test_ins_complete.vim
M  src/testdir/test_popup.vim

girishji committed:
Merge branch 'badselect' of https://github.com/girishji/vim into badselect

girishji committed:
Merge branch 'vim:master' into badselect

girishji committed:
Remove redundant code

M  src/insexpand.c

girishji committed:
Merge branch 'badselect' of https://github.com/girishji/vim into badselect
